cocos/core/assets/material.ts

```diff
@@ -441,32 +441,51 @@ export class Material extends Asset {
                 pass.resetUniform(name);
             }
         } else if (Array.isArray(val)) {
-            for (let i = 0; i < val.length; i++) {
-                this._bindTexture(pass, handle, val[i], i);
+            if (val.length !== 0) {
+                this._bindTextureArray(pass, name, val);
             }
         } else if (val) {
-            this._bindTexture(pass, handle, val);
+            this._bindTexture(pass, name, val);
         } else {
             pass.resetTexture(name);
         }
         return true;
     }
 
-    protected _bindTexture (pass: Pass, handle: number, val: MaterialPropertyFull, index?: number) {
-        const binding = Pass.getBindingFromHandle(handle);
+    protected _bindTexture (pass: Pass, name: string, val: MaterialPropertyFull) {
         if (val instanceof Texture) {
-            pass.bindTexture(binding, val, index);
+            pass.setTexture(name, val);
         } else if (val instanceof TextureBase) {
             const texture: Texture | null = val.getGFXTexture();
             if (!texture || !texture.width || !texture.height) {
                 // console.warn(`material '${this._uuid}' received incomplete texture asset '${val._uuid}'`);
                 return;
             }
-            pass.bindTexture(binding, texture, index);
-            pass.bindSampler(binding, val.getGFXSampler(), index);
+            pass.setTextureAndSampler(name, texture, val.getGFXSampler());
         }
     }
 
+    protected _bindTextureArray (pass: Pass, name: string, val: MaterialPropertyFull[]) {
+        const textures = new Array<Texture>(val.length);
+        const samplers = new Array<Sampler>(val.length);
```

Comment:
It will be triggered when properties are loaded or changed, like changing textures.
This should not happen frequently.
If a material need changing textures, user should consider using other rendering mechanism.